zst/object.go

```diff
@@ -59,36 +59,33 @@ func NewObject(zctx *zed.Context, s *storage.Seeker, size int64) (*Object, error
 	return o, nil
 }
 
-func NewObjectFromSeeker(zctx *zed.Context, s *storage.Seeker) (*Object, error) {
-	size, err := storage.Size(s.Reader)
+func NewObjectFromStorageReaderNoCloser(zctx *zed.Context, r storage.Reader) (*Object, error) {
+	size, err := storage.Size(r)
 	if err != nil {
 		return nil, err
 	}
-	return NewObject(zctx, s, size)
+	return NewObject(zctx, r, size)
 }
 
-func NewObjectFromPath(ctx context.Context, zctx *zed.Context, engine storage.Engine, path string) (*Object, error) {
-	uri, err := storage.ParseURI(path)
-	if err != nil {
-		return nil, err
-	}
-	r, err := engine.Get(ctx, uri)
+func NewObjectFromStorageReader(zctx *zed.Context, r storage.Reader) (*Object, error) {
+	o, err := NewObjectFromStorageReaderNoCloser(zctx, r)
 	if err != nil {
 		return nil, err
 	}
-	size, err := storage.Size(r)
+	o.closer = r.(io.Closer)
+	return o, nil
+}
+
+func NewObjectFromPath(ctx context.Context, zctx *zed.Context, engine storage.Engine, path string) (*Object, error) {
+	uri, err := storage.ParseURI(path)
 	if err != nil {
 		return nil, err
 	}
-	seeker, err := storage.NewSeeker(r)
+	r, err := engine.Get(ctx, uri)
 	if err != nil {
 		return nil, err
 	}
-	o, err := NewObject(zctx, seeker, size)
-	if err == nil {
-		o.closer = r
-	}
-	return o, err
+	return NewObjectFromStorageReader(zctx, r)
```

Comment:
We should close the storage.Reader on error.
```suggestion
	object, err := NewObjectFromStorageReader(zctx, r)
	if err != nil {
		r.Close()
		return nil, err
	}
	return object, nil
```